WordPressAuthenticator/OAuth/OAuthError.swift

```diff
@@ -0,0 +1,10 @@
+enum OAuthErrors {
```

Comment:
Using an `enum` means we can't accidentally call do `let something = OAuthErrors()`.

Reply:
The more I think about it, actually, the less I'm happy with `NSError`. Keen to hear what you think.

If / when we'll improve the error system, it might be good to do it in a dedicated PR.

Reply:
I've been thinking about this and I wonder if this would be good.

```swift
enum OAuthError {

  struct InconsistentASWebAuthenticationSessionCompletion: LocalizedError {
    let errorDescription = "ASWebAuthenticationSession authentication finished with neither a callback URL nor error"
  }

  struct OtherThingThatWentWrong: LocalizedError {}

  struct YetAnotherThingThatWentWrong: LocalizedError {
    let aProperty: Int
  }

  // ...
}
```

I've seen some codebases use `enum OAuthErro: Error` but I don't like that because you end up with long `switch case` for `errorDescription` and the other `LocalizedError` properties. This approach still has the nice namespacing, but also keeps each error self contained. Removing one would amount to remove a continuous group of lines, rather that a `case name` definition here and various `case .name: ...` usages there.

Reply:
For reference, [the Swift docs](https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html) use an example with `enum`:

```swift
enum VendingMachineError: Error {
    case invalidSelection
    case insufficientFunds(coinsNeeded: Int)
    case outOfStock
}
```

Another option I found which could be interesting is [using nested `enums`](https://vojtastavik.com/2017/04/23/organize-your-errors-swift/): 

```swift
enum Error: Swift.Error {
    case network(NetworkError)
    enum NetworkError {
        case notReachable
        case unknown
    }

    case device(DeviceError)
    enum DeviceError {
        case notEnoughSpace
        case unsupportedSystemVersion
    }
}
```

I like the idea of grouping all errors as `enum` cases, but I think I still prefer the encapsulation that `struct MyError: Error` brings. I would modify that example to:

```swift
protocol NetworkError: Swift.Error {}
protocol DeviceError: Swift.Error {}

enum Error: Swift.Error {
    case network(NetworkError)
    case device(DeviceError)

    enum Network {
        struct NotReachableError: NetworkError {}
        struct Unknown: NetworkError {}
    }

    enum Device: DeviceError {
        struct NotEnoughSpaceError: DeviceError {}
        struct UnsupportedSystemVersionError: DeviceError {}
    }
}

# Used in practice as

let e1 = Error.device(Error.Device.NotEnoughSpaceError())
let e2 = Error.network(Error.Network.NotReachableError())
```

Admittedly, one loses the ability to `switch` on the `NetworkError` and `DeviceError`, and it's unfortunate that the `NetworkError` and `DeviceError` protocols cannot be defined as nested in `enum Error`. Still, I think this would scale better.

Also, if really someone really wanted to switch between the leaves, they could do it via:

```swift
    switch error {
    case .network: break
    case .device(let deviceError):
        switch deviceError.self {
        case is Error.Device.UnsupportedSystemVersionError:
            print("Device error")
        case is Error.Device.NotEnoughSpaceError:
            print("Not enough space error")
        case _:
            fatalError("This code should not have been reached") // Yet another limitation of this approach
        }
    }
```

Reply:
In my experience, I only used error declaration similar to the `VendingMachineError` in your example. Yes, you do end up with a long switch case, but I think that's what you want in this instance, as you want _all_ errors to be handled, and Swift compiler can help you enforcing that (unless there is a `default` clause in the switch-case).

Regarding your second example, I don't think I have ever written a nested error type like. First of all, the associated value of the Error enum cases does not need to conform to `Swift.Error` (I don't see any benefit in doing that). Also, I feel like it's more readable to flatten the nested enum cases. In your example, there are really four possible errors (rather than two as declared in the `enum Error`): "network not reachable", unknown network issue, unsupported OS, and not enough disk space. I don't see the benefit of categorizing these four errors into "device" and "network".

Reply:
I've been trying out the `enum` collecting `struct E: LocalizedError` approach in my development branch, with [this](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/77fd7040680225cf92b0c12ec64569bcbca8bb0b/WordPressAuthenticator/OAuth/OAuthError.swift) being latest iteration at the time of writing.

I really dislike the spreading of information that we get with the `enum` approach, but I agree with the compile time benefit that you highlight.

> as you want all errors to be handled, and Swift compiler can help you enforcing that (unless there is a default clause in the switch-case).

None of the other approaches I tried is as secure and supported as what the compiler can do for us when using an `enum`.

As for nesting, leaving aside the oversight of having associated types not conforming to `Swift.Error`, I'd keep that as a technique up our sleeves to reduce the noise if we end up with a lot of errors, but I guess it might be best to have more specialized error types, e.g. `enum RequestError: Error` and `enum ResponseError: Error` that a catch-all error type with cases for both.